### PR TITLE
fix Issue 16589 - Taking address of stack variables in @safe code is …

### PR DESCRIPTION
**Pull request by WalterBright**
…allowed in some cases

### CODE REVIEW

**Comment by mathias-lang-sociomantic** [comment]
Should it be part of `-transition=safe` ?

**Comment by WalterBright** [comment]
Unfortunately, yes, as it is tripping in Phobos.

**Comment by WalterBright** [comment]
It's been an error for years to take the address of a local in `@safe` code. It does not make sense to allow taking the address of a field of a local. This was not checked for, and that's what this PR does. It is not part of DIP1000.

It does break at least one Phobos function (the autotester quits at the first error), which is why I amended it to only check if `-transition=safe`, but the code is incorrect. Allowing this in @safe code breaks the whole point of `ref`.

If the user needs to `&` a local, the way to do it is by `ref` instead, or make it `@trusted`.

BTW, this rule nicely resolves the ambiguity of what `return` refers to with a parameter marked as `return ref scope`. If the function returns by `ref`, the 'return' applies to the `ref`.

File: src/expression.d [hunk 1]
```diff
@@ -10581,6 +10581,28 @@ extern (C++) final class AddrExp : UnaExp
 
         type = e1.type.pointerTo();
 
+        bool checkAddressVar(VarDeclaration v)
+        {
+            if (v)
```
Comment (andralex) [inline src/expression.d]:
```d
if (!v)
    return true; // nuttin' to do
... rest of the code ...
```

File: src/expression.d [hunk 2]
```diff
@@ -10581,6 +10581,28 @@ extern (C++) final class AddrExp : UnaExp
...
+            {
+                if (!v.canTakeAddressOf())
+                {
+                    error("cannot take address of %s", e1.toChars());
+                    return false;
+                }
+                if (sc.func && !sc.intypeof && !v.isDataseg())
+                {
+                    if (sc.func.setUnsafe())
```
Comment (andralex) [inline src/expression.d]: `if` + `if` = `&&`

```d
if (sc.func && !sc.intypeof && !v.isDataseg() && sc.func.setUnsafe())
{
    ...
}
```

File: src/expression.d [hunk 3]
```diff
@@ -10604,26 +10626,36 @@ extern (C++) final class AddrExp : UnaExp
...
             if (checkUnsafeAccess(sc, dve, !type.isMutable(), true))
                 return new ErrorExp();
+
+            if (dve.e1.op == TOKvar && global.params.safe)
+            {
+                VarExp ve = cast(VarExp)dve.e1;
+                VarDeclaration v = ve.var.isVarDeclaration();
+                if (v)
+                {
+                    if (!checkAddressVar(v))
```
Comment (andralex) [inline src/expression.d]:
```d
if (v && !checkAddressVar(v))
    return new ErrorExp();
```

File: src/expression.d [hunk 4]
```diff
@@ -10604,26 +10626,36 @@ extern (C++) final class AddrExp : UnaExp
...
+                        return new ErrorExp();
+                }
+            }
+            else if ((dve.e1.op == TOKthis || dve.e1.op == TOKsuper) && global.params.safe)
+            {
+                ThisExp ve = cast(ThisExp)dve.e1;
+                VarDeclaration v = ve.var.isVarDeclaration();
+                if (v && v.storage_class & STCref)
+                {
+                    if (!checkAddressVar(v))
```
Comment (andralex) [inline src/expression.d]:
```d
if (v && v.storage_class & STCref) && !checkAddressVar(v))
    return new ErrorExp();
```

File: src/expression.d [hunk 5]
```diff
@@ -10675,6 +10707,16 @@ extern (C++) final class AddrExp : UnaExp
                 }
             }
         }
+        else if ((e1.op == TOKthis || e1.op == TOKsuper) && global.params.safe)
+        {
+            ThisExp ve = cast(ThisExp)e1;
+            VarDeclaration v = ve.var.isVarDeclaration();
+            if (v)
+            {
+                if (!checkAddressVar(v))
```
Comment (andralex) [inline src/expression.d]:
```d
if (v && !checkAddressVar(v))
    return new ErrorExp();
```

**Review by andralex — APPROVED** [review]
Few nits to improve the lot of readers and reduce LOC.

**Comment by WalterBright** [comment]
Auto-merge toggled on

**Pull request merged**